internal/privatestate/data_test.go

```diff
@@ -0,0 +1,307 @@
+package privatestate_test
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/internal/privatestate"
+)
+
+func TestData_Bytes(t *testing.T) {
+	t.Parallel()
+
+	empty, _ := json.Marshal(privatestate.Data{})
+
+	frameworkData, _ := json.Marshal(privatestate.Data{
+		Framework: map[string][]byte{
+			"frameworkKeyOne": []byte("framework value one"),
+			"frameworkKeyTwo": []byte("framework value two"),
+		},
+	})
+
+	providerData, _ := json.Marshal(privatestate.Data{
+		Provider: map[string][]byte{
+			"providerKeyOne": []byte("provider value one"),
+			"providerKeyTwo": []byte("provider value two"),
+		},
+	})
+
+	frameworkProviderData, _ := json.Marshal(privatestate.Data{
+		Framework: map[string][]byte{
+			"frameworkKeyOne": []byte("framework value one"),
+			"frameworkKeyTwo": []byte("framework value two"),
+		},
+		Provider: map[string][]byte{
+			"providerKeyOne": []byte("provider value one"),
+			"providerKeyTwo": []byte("provider value two"),
+		},
+	})
+
+	testCases := map[string]struct {
+		data          privatestate.Data
+		expected      []byte
+		expectedDiags diag.Diagnostics
```

Comment:
It'd be great to have a test case that returns an error diagnostic, maybe something like bytes that don't successfully convert to a valid JSON string?

Reply:
Believe we'll need to introduce a check in the `Bytes` func that determines whether each of the `[]byte` that is associated with each key is valid JSON (i.e., `json.Valid`) in order to be able to add that test. Is that what you had in mind?

Reply:
Ah ha, you can technically convince `json.Marshal()` itself to error on invalid map key types, e.g., https://go.dev/play/p/CFA4MqvyoPw

```go
json.Marshal(
	map[interface{}]struct{}{
		big.NewFloat(123): {},
	},
)
```

However our usage of `map[string][]byte` for storage prevents that particular issue. I think we're good here. 😄